---
title: "Pipeline of Differential Expression Analysis of KIRC data"
output: 
  html_document: 
    default
  github_document: 
    df_print: paged
    html_preview: FALSE
    keep_html: TRUE
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_format = "all") })    
---

Full pipeline of DEA with Glimma - Comparing gene expression by sample type of KIRC - Kidney Renal Clear Cell Carcinoma

https://www.cbioportal.org/study/summary?id=kirc_tcga_pub


```{r, error=TRUE, message=FALSE, warning=FALSE, purl=FALSE, results='hide'}
## This chunk automatically generates a text .R version of this script when running within knitr.
input  = knitr::current_input()  # filename of input document
output = paste(tools::file_path_sans_ext(input), 'R', sep = '.')
try(knitr::purl(input,output,documentation=2,quiet=T), silent = T)
# Avoid duplicate label error of knitr::purl
options(knitr.duplicate.label = 'allow')
# Code to browse the markdown file with renderized images.
knitr::opts_chunk$set(
  fig.path = "figs/render-"
)
```



# Installing and Loading Libraries            


```{r message=FALSE, warning=FALSE, paged.print=FALSE}
packages_bioconductor = c("TCGAbiolinks", "SummarizedExperiment", "DESeq2", "Glimma", "limma","biomaRt", "genefilter")

#use this function to check if each package is on the local machine
#if a package is installed, it will be loaded
#if any are not, the missing package(s) will be installed from Bioconductor and loaded
package.check <- lapply(packages_bioconductor, FUN = function(x) {
  if (!require(x, character.only = TRUE)) {
    BiocManager::install(x, dependencies = TRUE)
    library(x, character.only = TRUE)
  }
})

packages_cran = c("DT", "tidyverse","dplyr", "tibble", "stringr", "data.table", "genefilter", "ggrepel")
  
#use this function to check if each package is on the local machine
#if a package is installed, it will be loaded
#if any are not, the missing package(s) will be installed from CRAN and loaded
package.check <- lapply(packages_cran, FUN = function(x) {
  if (!require(x, character.only = TRUE)) {
    install.packages(x, dependencies = TRUE)
    library(x, character.only = TRUE)
  }
})

rm(packages_cran, packages_bioconductor, package.check)
```



<!-- # Download - Expression Data                   -->



<!-- ```{r} -->
<!-- query.exp.KIRC <- GDCquery(project = "TCGA-KIRC", -->
<!--                            legacy = TRUE, -->
<!--                            data.category = "Gene expression", -->
<!--                            data.type="Gene expression quantification", -->
<!--                            file.type="results", # not normalized! -->
<!--                            platform = "Illumina HiSeq", -->
<!--                            experimental.strategy = "RNA-Seq") -->

<!-- GDCdownload(query.exp.KIRC,files.per.chunk = 100) -->

<!-- query.exp.KIRC <- GDCprepare(query = query.exp.KIRC, -->
<!--                              save = TRUE, -->
<!--                              save.filename = "qquery.exp.KIRC0.rda" , -->
<!--                              summarizedExperiment = TRUE) -->
<!-- ``` -->



<!-- <!-- # Download - Mutation Data   --> -->


<!-- #```{r} -->
<!-- #KIRC_maf <- GDCquery_Maf("KIRC", pipelines = "muse") -->
<!-- #``` -->



<!-- ```{r} -->
<!-- load("~/bio/josivan_bimodal_genes/dados_SD_TP.RData") -->
<!-- rm(KIRC_CONT_SD, KIRC_CONT_TP) -->

<!-- kirc$V2 <- NULL -->
<!-- kirc$SampleType <- NULL -->

<!-- kirc <- kirc[!duplicated(kirc[,c("V1","V3")]),] -->

<!-- exp <- kirc %>%  -->
<!--             tidyr::spread(V1, V4) -->

<!-- # exp_count <- pivot_wider( -->
<!-- #   kirc, -->
<!-- #   names_from = V1, -->
<!-- #   values_from = V4) -->

<!-- row.names(exp) <- exp$V3 -->
<!-- exp$V3 <- NULL -->
<!-- exp <- as.matrix(exp) -->
<!-- rm(kirc) -->
<!-- ``` -->



<!-- ```{r} -->
<!-- load("~/Dropbox/bioinformatics/research/bimodal_genes/DEA/qquery.exp.KIRC0.rda") -->
<!-- ``` -->

<!-- # Harmonizating Expression and clinical Data   -->


<!-- ```{r} -->

<!-- #exp <- assays(data)[[1]] -->
<!-- #rm(query.exp.KIRC) -->

<!-- data_description <- data.frame(codes=data$barcode,  -->
<!--                                vital_status=data$vital_status,  -->
<!--                                sample_type=data$sample_type, -->
<!--                                metastasis=data$ajcc_pathologic_m) -->


<!-- data_description$codes <- substr(data_description$codes,1,16) -->

<!-- data_description <- data_description %>%  -->
<!--    mutate(sample_type = recode(sample_type,  -->
<!--                                 "Solid Tissue Normal" = "normal", -->
<!--                                 "Primary Tumor" = "tumor")) %>% -->
<!--    filter(sample_type %in% c("normal", "tumor")) %>% droplevels() -->

<!-- #data_description <- data_description %>%  -->
<!-- #  filter(metastasis %in% c("M0", "M1")) %>% droplevels() -->

<!-- data_description$sample_type <- factor(as.character(data_description$sample_type), levels = c("normal", "tumor")) -->
<!-- #data_description$metastasis <- factor(as.character(data_description$metastasis), levels = c("M0", "M1")) -->


<!-- row.names(data_description) <- data_description$codes -->

<!-- exp <- as.matrix(exp[,colnames(exp) %in% rownames(data_description)]) -->

<!-- # Converting the expression data to integer -->
<!-- #storage.mode(exp) = "integer" -->

<!-- #rownames(exp) <- str_split_fixed(rownames(exp), '[|]', 2)[,1] -->
<!-- #exp <- as.matrix(exp[!duplicated(row.names(exp)), ]) -->

<!-- #colnames(exp) <- row.names(data_description)  -->
<!-- data_description <- data_description[colnames(exp),] -->
<!-- table(data_description$sample_type) -->
<!-- table(data_description$metastasis) -->

<!-- rm(data) -->
<!-- ``` -->


<!-- # Select BioMart database and dataset         -->


<!-- ```{r} -->
<!-- # If you haven't already installed devtools... -->
<!-- #install.packages("devtools") -->

<!-- # Use devtools to install the package -->
<!-- #devtools::install_github("stephenturner/annotables") -->

<!-- library(annotables) -->

<!-- ensemblAnnot <- grch38 %>%  -->
<!--   filter(grch38$symbol %in%  row.names(exp))  %>%  -->
<!--    dplyr::select(ensgene, symbol, description) -->

<!-- colnames(ensemblAnnot) <- c("Ensembl", "GeneID", "Description") -->

<!-- #genes_miss <- generics::setdiff(row.names(exp), unique(ensemblAnnot$Symbol)) -->


<!-- ``` -->


<!-- # Alterative way to get gene Annotations by biomart package -->

<!-- ```{r} -->

<!-- ennsembl = useMart("ensembl",dataset="hsapiens_gene_ensembl")  -->

<!-- ensembl_genes <- row.names(exp)  -->

<!-- ensemblAnnot <- getBM(attributes= c("hgnc_symbol", "ensembl_gene_id", "description"),  -->
<!--                                           filters=c("hgnc_symbol"),  -->
<!--                                           values=ensembl_genes, mart=ensembl) -->


<!-- genes_miss <- setdiff(ensembl_genes, unique(ensemblAnnot$hgnc_symbol)) -->

<!-- while(length(genes_miss) > 10){ -->

<!-- for(i in seq(from=1, to=length(genes_miss), by=100)){ -->
<!--    try(ensemblAnnot_1 <- getBM(attributes= c("hgnc_symbol", "ensembl_gene_id", "description"),  -->
<!--                                               filters=c("hgnc_symbol"),  -->
<!--                                               values=genes_miss[i:(i+99)], mart=ensembl), silent = T) -->
<!--    ensemblAnnot <- rbind(ensemblAnnot, ensemblAnnot_1) -->
<!--    ensemblAnnot <- ensemblAnnot[!duplicated(ensemblAnnot),] -->
<!--    genes_miss <- setdiff(ensembl_genes, unique(ensemblAnnot$hgnc_symbol)) -->
<!--    print(i) -->
<!--    Sys.sleep(1) -->
<!-- } -->
<!--   genes_miss <- setdiff(ensembl_genes, unique(ensemblAnnot$hgnc_symbol)) -->
<!-- } -->

<!-- colnames(ensemblAnnot) <- c("GeneID", "Ensembl", "Description")  -->

<!-- rm(ensembl,ensembl_genes) -->
<!-- ``` -->


<!-- # Alterative way to get gene Annotations by mygene package -->

<!-- ```{r} -->
<!-- library(mygene) -->
<!-- out <- queryMany(ensembl_genes, scopes="symbol", fields=c("symbol", "ensembl.gene", "description"), species="human") -->
<!-- ``` -->


```{r}
data_description <- kirp_clin %>% select(  
                                            )
     
```
```{r}

 row.names(data_description) <- data_description$codes

 exp <- as.matrix(exp[,colnames(exp) %in% rownames(data_description)])

```



```{r}
data_description <- kirp_clin %>% select(c(  'Study ID', 
                                              'Overall Survival Status',
                                              'Sample Type',
                                              'Prior Diagnosis'
                                            )
                                         )
```


```{r}
#### kirp_clin <-  data_description %>% 
####              filter('prior_diagnosis' %in% c("No", "Yes","Yes, History Of 
#### Synchronous And Or Bilateral Malignancy")) %>% 
####              droplevels()

exp_sample_type <- kirp_clin %>% select(c('Study ID'))
data_sample_type <- kirp_clin %>% select(c('Sample Type'))
prior_diagnosis <- kirp_clin %>% select(c('Prior Diagnosis'))
```

# Running DESeq2 Differential Expression


```{r}

# create the DESeqDataSet object
ddsObj <- DESeqDataSetFromMatrix(countData = exp_sample_type,
                              colData = data_sample_type,
                              design = ~ prior_diagnosis)
```
    

# Normalisation


```{r}
# Apply normalisation to DDS object
ddsObj <- estimateSizeFactors(ddsObj)
```
    
    
# Interactive StripChart with Glimma
    
    
```{r}
ddsObj <- DESeq(ddsObj)
res.DESeq2 <- results(ddsObj)
res.shr <- DESeq2::lfcShrink(ddsObj, coef=2, res=res.DESeq2)

```
```{r}
summary(res.DESeq2)

```

```{r}
summary(res.shr)

```

```{r}
shrink.deseq <- as.data.frame(res.DESeq2) %>%
    rownames_to_column("symbol") %>% 
    left_join(annotation, by="symbol") %>% 
    rename(logFC=log2FoldChange, FDR=padj)
```



```{r}
df.deseq <- shrink.deseq[abs(shrink.deseq$logFC) > 1 & shrink.deseq$pvalue < 0.01,]

df.deseq <- df.deseq[!is.na(df.deseq$logFC),]
dim(df.deseq)
```


```{r}
genes.DEA.NT.vs.TP.lst <- unique(df.deseq$symbol)
genes.DEA.NT.vs.TP.lst <- genes.DEA.NT.vs.TP.lst[!grepl('^LOC', genes.DEA.NT.vs.TP.lst)]
write(genes.DEA.NT.vs.TP.lst,  file = "genes.DEA.NT.vs.TP.lst")
```


<!-- ```{r} -->
<!-- DESeq2::plotMA(res, ylim=c(-5,5)) -->
<!-- ``` -->



<!-- ```{r} -->

<!-- # first remove the filtered genes (FDR=NA) and create a -log10(FDR) column -->
<!-- filtTab.deseq <- shrink.deseq %>% -->
<!--     filter(!is.na(FDR)) %>% -->
<!--     mutate(`-log10(FDR)` = -log10(FDR)) -->

<!--   filtTab.deseq <- filtTab.deseq  %>%  -->
<!--     mutate(`-log10(FDR)`=pmin(`-log10(FDR)`)) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- res.df <- as.data.frame(res.shr) -->
<!-- res.df$log10MeanNormCount <- log10(res.df$baseMean + 1) -->
<!-- idx <-(rowSums(counts(ddsObj)) > 0) -->
<!-- res.df <- res.df[idx,] -->
<!-- res.df$padj[is.na(res.df$padj)] <- 1 -->

<!-- status <- as.numeric(res.df$padj < .1) -->

<!-- glMDPlot(res.df[idx,], -->
<!--          xval="baseMean", -->
<!--          yval="log2FoldChange", -->
<!--          counts=counts(ddsObj)[idx,], -->
<!--          anno=data.frame(GeneID=rownames(ddsObj)[idx]), -->
<!--          groups=ddsObj$sample_type, -->
<!--          samples=colnames(ddsObj), -->
<!--          status=status, -->
<!--          display.columns=c("GeneID"), -->
<!--          folder = "volcano_KIRC_deseq_DE", -->
<!--          launch=FALSE) -->

<!-- ``` -->


<!-- ```{r} -->
<!-- de <- as.integer(filtTab.deseq$FDR <= 0.05) -->

<!-- normCounts <- log2(counts(ddsObj)) -->
<!-- filtCounts <- normCounts[filtTab.deseq$GeneID,] -->

<!-- glXYPlot( -->
<!--   x = filtTab.deseq$logFC, -->
<!--   y = -log10(filtTab.deseq$FDR), -->
<!--   xlab = "logFC", -->
<!--   ylab = "FDR", -->
<!--   main = "Normal.vs.Tumor", -->
<!--   counts = filtCounts, -->
<!--   groups = data_description$sample_type, -->
<!--   status = de, -->
<!--   anno = filtTab.deseq[, c("GeneID", "Ensembl", "Description")], -->
<!--   folder = "volcano_KIRC_deseq_DE", -->
<!--   launch = F -->
<!-- ) -->
<!-- ``` -->

```{r}

#save.image("KIRC_pipeline_glimma_DEA_count.RData")

```


<!-- ```{r} -->
<!-- if(!require("RColorBrewer")){install.packages("RColorBrewer")} -->
<!-- if(!require("circlize")){install.packages("circlize")} -->


<!-- # get the top genes -->
<!-- sigGenes <- as.data.frame(res.shr) %>%  -->
<!--     rownames_to_column("GeneID") %>%  -->
<!--     top_n(150, wt=-padj) %>%  -->
<!--     pull("GeneID") -->
<!-- # filter the data for the top 200 by padj in the LRT test -->
<!-- plotDat <- vst(ddsObj)[sigGenes,] %>%  -->
<!--     assay() -->
<!-- z.mat <- t(scale(t(plotDat), center=TRUE, scale=TRUE)) -->

<!-- # colour palette -->
<!-- myPalette <- c("red3", "ivory", "blue3") -->
<!-- myRamp = colorRamp2(c(-2, 0, 2), myPalette) -->
<!-- ``` -->



<!-- ```{r} -->
<!-- if(!require("ComplexHeatmap")){install.packages("ComplexHeatmap")} -->

<!-- # cluster the data and split the tree -->
<!-- hcDat <- hclust(dist(z.mat)) -->
<!-- cutGroups <- cutree(hcDat, h=2) -->
<!-- ha1 = HeatmapAnnotation(df = colData(ddsObj)[,c("sample_type", "vital_status")]) -->
<!-- Heatmap(z.mat, name = "z-score", -->
<!--         col = myRamp,             -->
<!--         show_row_name = FALSE, -->
<!--         cluster_columns = FALSE, -->
<!--         split=cutGroups, -->
<!--         rect_gp = gpar(col = "darkgrey", lwd=0.5), -->
<!--         top_annotation = ha1, -->
<!--         row_names_gp = gpar(fontsize = 8), -->
<!--         column_names_gp = gpar(fontsize = 8), -->
<!--         column_names_rot = 45) -->

<!-- rm(hcDat,cutGroups, ha1 ) -->
<!-- ``` -->


<!-- # Constructing the DEA test with Limma -->
<!-- ```{r} -->
<!-- vm <- voomWithQualityWeights(eset, design=design) -->
<!-- fit <- lmFit(vm, design=design) -->
<!-- fit <- contrasts.fit(fit, contrasts) -->
<!-- fit <- eBayes(fit) -->
<!-- dt <- decideTests(fit) -->
<!-- DE <- c("downregulated", "notDE", "upregulated")[as.factor(dt)] -->
<!-- anno <- as.data.frame(cbind(data.frame(GeneID=rownames(fit)), DE)) -->

<!-- anno <- as.data.frame(cbind(anno, ensemblAnnot[anno$GeneID, c("Ensembl", "Description")])) -->
<!-- rownames(anno) <- anno$GeneID -->

<!-- ``` -->

<!-- # Creating labels for "downregulated", "notDE" and "upregulated" genes -->
<!-- ```{r} -->

<!-- #head(anno) -->

<!-- filtTab.limma <- shrink.limma %>%  -->
<!--   filter(!is.na(FDR)) %>%  -->
<!--   mutate(`-log10(FDR)` = -log10(FDR)) %>%  -->
<!--   mutate(TopGeneLabel=ifelse(pvalue<=cutoff, Ensembl, "")) -->

<!-- de <- as.integer(filtTab.limma$FDR[anno$GeneID] <= 0.05) -->

<!-- filtCounts <- exp[anno$GeneID,] -->

<!-- logFC <- filtTab.limma$logFC[filtTab.limma$GeneID %in% anno$GeneID] -->

<!-- ``` -->

<!-- # Creating the HTML with Glimma -->


<!-- ```{r} -->

<!-- cols <- c("yellow","blue","magenta") -->

<!-- sample.cols <- c("limegreen", "purple")[groups] -->

<!-- glXYPlot(x=fit$coef, y=fit$lod, xlab="logFC", ylab="logodds",status=dt, anno=anno, side.main="GeneID",counts=vm, groups=groups, sample.cols=sample.cols, -->
<!--          folder = "volcano_KIRC_limma_DE" -->
<!--          ) -->
<!-- ``` -->
